crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+            .map_err(|e| StorageError::Internal(format!("GSI KeySchema build error: {e}")))?;
+
+        let gsi_range_key = KeySchemaElement::builder()
+            .attribute_name("relay_after")
+            .key_type(KeyType::Range)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("GSI KeySchema build error: {e}")))?;
+
+        let throughput = ProvisionedThroughput::builder()
+            .read_capacity_units(5)
+            .write_capacity_units(5)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("Throughput build error: {e}")))?;
+
+        let gsi_throughput = ProvisionedThroughput::builder()
+            .read_capacity_units(5)
+            .write_capacity_units(5)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("GSI Throughput build error: {e}")))?;
+
+        let projection = Projection::builder()
+            .projection_type(ProjectionType::All)
+            .build();
+
+        let gsi = GlobalSecondaryIndex::builder()
+            .index_name(STATUS_RELAY_INDEX)
+            .key_schema(gsi_hash_key)
+            .key_schema(gsi_range_key)
+            .projection(projection)
+            .provisioned_throughput(gsi_throughput)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("GSI build error: {e}")))?;
+
+        if let Err(e) = self
+            .client
+            .create_table()
+            .table_name(TABLE_NAME)
+            .attribute_definitions(attr_def_id)
+            .attribute_definitions(attr_def_status)
+            .attribute_definitions(attr_def_relay_after)
+            .key_schema(key_schema)
+            .global_secondary_indexes(gsi)
+            .provisioned_throughput(throughput)
+            .send()
+            .await
+        {
+            return Err(StorageError::Internal(format!(
+                "Failed to create DynamoDB table: {e}"
+            )));
+        }
+
+        Ok(())
+    }
+
+    async fn query_requests_by_status_and_time(
+        &self,
+        status: RequestStatus,
+        max_timestamp: &str,
+        limit: usize,
+    ) -> Result<Vec<ScheduledRequest>, StorageError> {
+        let mut requests = Vec::new();
+        let mut last_evaluated_key = None;
+
+        loop {
+            let mut query_builder = self
+                .client
+                .query()
+                .index_name(STATUS_RELAY_INDEX)
+                .key_condition_expression("#status = :status AND relay_after <= :max_time")
+                .expression_attribute_names("#status", "status")
+                .expression_attribute_values(
+                    ":status",
+                    AttributeValue::S(status_to_str(&status).to_string()),
+                )
+                .expression_attribute_values(
+                    ":max_time",
+                    AttributeValue::N(max_timestamp.to_string()),
+                )
+                .limit((limit - requests.len()) as i32);
```

Comment:
Not sure if to use that limit optimization. First of all, whether to use `limit` at all. DynamoDB has its "hardware" limit to request up to 1 MB of data. So I'd either remove this line, or remove subtraction. I am worried this subtraction with the combination of `loop` can return too few results in some scenarios. 

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.Other.html